src/core/Akka/Actor/GracefulStopSupport.cs

```diff
@@ -63,19 +63,25 @@ public static Task<bool> GracefulStop(this IActorRef target, TimeSpan timeout, o
             return promise.Task.ContinueWith(t =>
             {
                 var returnResult = false;
-                PatternMatch.Match(t.Result)
-                    .With<Terminated>(terminated =>
-                    {
-                        returnResult = (terminated.ActorRef.Path.Equals(target.Path));
-                    })
-                    .Default(m =>
-                    {
-                        returnResult = false;
-                    });
-
-                internalTarget.Tell(new Unwatch(target, fref));
+                if (t.IsCompleted)
+                {
+                    PatternMatch.Match(t.Result)
+                        .With<Terminated>(terminated =>
+                        {
+                            returnResult = (terminated.ActorRef.Path.Equals(target.Path));
+                        })
+                        .Default(m =>
+                        {
+                            internalTarget.Tell(new Unwatch(target, fref));
+                            returnResult = false;
+                        });
+                }
+                else //need to Unwatch in the event of a timeout or failure
+                {
```

Comment:
No race conditions (because this occurs inside a continuation) but I just realized that `Task.IsCompleted` is the wrong flag to check.